rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py

```diff
@@ -1,203 +1,135 @@
-                1
-            ].sample()
-
-        else:
-            # Get the posterior action distribution to sample the posterior action.
-            posterior_action_dist = self.action_dist_cls.from_logits(posterior_logits)
-            # Sample the posterior action.
-            posterior_action = posterior_action_dist._flat_child_distributions[
-                1
-            ].sample()
-            # We need the log-probabilities for the loss.
-            pi_outs[Columns.ACTION_LOGP] = posterior_action_dist.logp(
-                (prior_action, posterior_action)
-            )
-            # We also need the input to the action distribution to calculate the
-            # KL-divergence.
-            pi_outs[Columns.ACTION_DIST_INPUTS] = posterior_logits
-
-        # Concatenate the prior and posterior actions and log probabilities.
-        pi_outs[Columns.ACTIONS] = (prior_action, posterior_action)
-
-        return pi_outs
-
     @override(TorchRLModule)
     def _forward_inference(self, batch: Dict[str, TensorType]) -> Dict[str, TensorType]:
-        # Encoder forward pass.
-        encoder_out = self.encoder(batch)
-        # Policy head forward pass.
-        return self.pi(encoder_out[ENCODER_OUT], inference=True)
+        return self._pi(batch[Columns.OBS], inference=True)
 
     @override(TorchRLModule)
     def _forward_exploration(
         self, batch: Dict[str, TensorType], **kwargs
     ) -> Dict[str, TensorType]:
-        # Encoder forward pass.
-        encoder_out = self.encoder(batch)
-        # Policy head forward pass.
-        return self.pi(encoder_out[ENCODER_OUT], inference=False)
+        return self._pi(batch[Columns.OBS], inference=False)
 
     @override(TorchRLModule)
     def _forward_train(self, batch: Dict[str, TensorType]) -> Dict[str, TensorType]:
         return self._forward_exploration(batch)
 
     @override(ValueFunctionAPI)
     def compute_values(self, batch: Dict[str, TensorType], embeddings=None):
-        # Encoder forward pass to get `embeddings`, if necessary.
-        if embeddings is None:
-            embeddings = self.encoder(batch)[ENCODER_OUT]
-        # Value head forward pass.
-        vf_out = self.vf(embeddings)
+        # Value function forward pass.
+        vf_out = self._value_net(batch[Columns.OBS])
         # Squeeze out last dimension (single node value head).
         return vf_out.squeeze(-1)
+
+    def _pi(self, obs, inference: bool):
+        # Prior forward pass.
+        prior_out = self._prior_net(obs)
+        dist_a1 = TorchCategorical.from_logits(prior_out)
+
+        # If in inference mode, we need to set the distribution to be deterministic.
+        if inference:
+            dist_a1 = dist_a1.to_deterministic()
+        # Sample a1.
+        a1 = dist_a1.sample()
+
+        # Posterior forward pass.
+        posterior_batch = torch.cat(
+            [obs, one_hot(a1, self.action_space[0])],
```

Comment:
Interesting. I tried in multiple examples both, `one-hot` and "raw" actions and could not see a difference. HOw many iterations does it need now to converge?

Reply:
Yeah, I think the existing example was not learning or just super slowly. The new example definitely learns both action components' dependency on 1) obs and 2) obs+[first action].